File: src/api/commands.ts
// Typed wrappers over Tauri invoke(). The only place where the frontend knows command names.
import { Channel, invoke as tauriInvoke } from "@tauri-apps/api/core";
import { mockAiChat, mockInvoke } from "./mock";
import type {
  AiChatRequest,
  AiEndpoint,
  AiEvent,
  AiModel,
  ApplyResult,
  ColumnInfo,
  ConnectionConfig,
  ConnectionInput,
  CountRequest,
  ExecuteRequest,
  ExportRequest,
  ExportSummary,
  ForeignKeyInfo,
  IndexInfo,
  KeyListing,
  ParamStatement,
  QueryHistoryEntry,
  RowsExportRequest,
  ServerInfo,
  StatementResult,
  TableInfo,
} from "./types";

/** In a browser without Tauri (UI development), commands are served by the mock. */
const isTauri = typeof window !== "undefined" && "__TAURI_INTERNALS__" in window;
const invoke: typeof tauriInvoke = isTauri
  ? tauriInvoke
  : (cmd, args) => mockInvoke(cmd, args as Record<string, unknown>);

// --- Connections -------------------------------------------------------------

export const listConnections = () => invoke<ConnectionConfig[]>("list_connections");

export const saveConnection = (input: ConnectionInput) => invoke<ConnectionConfig>("save_connection", { input });

export const deleteConnection = (id: string) => invoke<void>("delete_connection", { id });

/** Tests a connection without saving it. */
export const testConnection = (input: ConnectionInput) => invoke<ServerInfo>("test_connection", { input });

/** Opens a connection pool for the connection (required before any queries). */
export const connect = (connectionId: string) => invoke<ServerInfo>("connect", { connectionId });

/** Closes the pool and all sessions of the connection. */
export const disconnect = (connectionId: string) => invoke<void>("disconnect", { connectionId });

// --- Schema --------------------------------------------------------------

export const listDatabases = (connectionId: string) => invoke<string[]>("list_databases", { connectionId });

export const listTables = (connectionId: string, database: string) =>
  invoke<TableInfo[]>("list_tables", { connectionId, database });

/** Keys of a key-value engine matching a glob pattern (`*` for all), at most `limit` of them. */
export const listKeys = (connectionId: string, database: string, pattern: string, limit: number) =>
  invoke<KeyListing>("list_keys", { connectionId, database, pattern, limit });

export const listColumns = (connectionId: string, database: string, table: string) =>
  invoke<ColumnInfo[]>("list_columns", { connectionId, database, table });

export const listIndexes = (connectionId: string, database: string, table: string) =>
  invoke<IndexInfo[]>("list_indexes", { connectionId, database, table });

export const listForeignKeys = (connectionId: string, database: string, table: string) =>
  invoke<ForeignKeyInfo[]>("list_foreign_keys", { connectionId, database, table });

export const getTableDdl = (connectionId: string, database: string, table: string) =>
  invoke<string>("get_table_ddl", { connectionId, database, table });

// --- Execution -------------------------------------------------------------

/** Splits sql into statements and runs them sequentially in the session. */
export const executeQuery = (request: ExecuteRequest) => invoke<StatementResult[]>("execute_query", { request });

/** Writes the complete result of a statement (no row limit) to a file on the backend. */
export const exportQuery = (request: ExportRequest) => invoke<ExportSummary>("export_query", { request });

/** Writes the given rows (the grid contents) to a file on the backend. */
export const exportRows = (request: RowsExportRequest) => invoke<ExportSummary>("export_rows", { request });

/** Counts the rows a statement produces (SELECT COUNT(*) over it). */
export const countQuery = (request: CountRequest) => invoke<number>("count_query", { request });

/** KILL QUERY for the query started with this queryId. */
export const cancelQuery = (connectionId: string, queryId: string) =>
  invoke<void>("cancel_query", { connectionId, queryId });

/** Runs parameterized statements in a single transaction (data editing). */
export const applyChanges = (connectionId: string, sessionId: string, statements: ParamStatement[]) =>
  invoke<ApplyResult>("apply_changes", { connectionId, sessionId, statements });

/** Closes the session connection (when a tab is closed). */
export const closeSession = (connectionId: string, sessionId: string) =>
  invoke<void>("close_session", { connectionId, sessionId });

// --- History ---------------------------------------------------------------

export const listHistory = (limit = 200) => invoke<QueryHistoryEntry[]>("list_history", { limit });

export const clearHistory = () => invoke<void>("clear_history");

// --- AI assistant ------------------------------------------------------------

/** Which of the given provider ids have a saved API key in the backend's SecretStore. */
export const aiKeyStatus = (providerIds: string[]) => invoke<Record<string, boolean>>("ai_key_status", { providerIds });

/** Saves an API key for a provider; an empty string deletes it. */
export const aiSetKey = (providerId: string, apiKey: string) => invoke<void>("ai_set_key", { providerId, apiKey });

export const aiDeleteKey = (providerId: string) => invoke<void>("ai_delete_key", { providerId });

/** Lists a provider's models; also doubles as "verify this API key" for the Settings dialog. */
export const aiListModels = (endpoint: AiEndpoint) => invoke<AiModel[]>("ai_list_models", { endpoint });

/**
 * Streams a chat completion, calling `onEvent` for every delta/done event and resolving with the
 * full response text once the model finishes (or rejecting, e.g. with "Cancelled" after `aiCancel`).
 * In the browser mock (no Tauri runtime) a fake streamer is used instead of `Channel`, which needs
 * the real Tauri IPC internals and cannot be constructed outside a Tauri webview.
 */
export const aiChat = (request: AiChatRequest, onEvent: (event: AiEvent) => void): Promise<string> => {
  if (!isTauri) return mockAiChat(request, onEvent);
  const channel = new Channel<AiEvent>();
  channel.onmessage = onEvent;
  return tauriInvoke<string>("ai_chat", { request, onEvent: channel });
};

/** Aborts a running `aiChat` by its requestId; no error if it is already finished or unknown. */
export const aiCancel = (requestId: string) => invoke<void>("ai_cancel", { requestId });

File: src-tauri/src/commands.rs
//! Thin `#[tauri::command]` wrappers — one-to-one with the list in `src/api/commands.ts`.

use tauri::State;

use crate::connections::{ConnectionConfig, ConnectionInput, ConnectionStore};
use crate::db::execute::{self, ExecuteRequest, ExportRequest};
use crate::db::export::{self, ExportSummary, RowsExportRequest};
use crate::db::{
    ApplyResult, ColumnInfo, ConnectionManager, ForeignKeyInfo, IndexInfo, ParamStatement, ServerInfo, StatementResult,
    TableInfo,
};
use crate::error::AppResult;
use crate::history::{History, QueryHistoryEntry};

pub struct AppState {
    pub connections: ConnectionStore,
    pub manager: ConnectionManager,
    pub history: History,
}

// --- Connections -----------------------------------------------------------

#[tauri::command]
pub async fn list_connections(state: State<'_, AppState>) -> AppResult<Vec<ConnectionConfig>> {
    state.connections.list()
}

#[tauri::command]
pub async fn save_connection(state: State<'_, AppState>, input: ConnectionInput) -> AppResult<ConnectionConfig> {
    state.connections.save(input)
}

#[tauri::command]
pub async fn delete_connection(state: State<'_, AppState>, id: String) -> AppResult<()> {
    state.connections.delete(&id)
}

/// Tests a connection without saving it.
#[tauri::command]
pub async fn test_connection(input: ConnectionInput) -> AppResult<ServerInfo> {
    ConnectionManager::test_connection(&input.to_view(), input.credentials()).await
}

/// Opens the driver for a connection (required before any queries).
#[tauri::command]
pub async fn connect(state: State<'_, AppState>, connection_id: String) -> AppResult<ServerInfo> {
    let config = state.connections.get_stored(&connection_id)?;
    let credentials = state.connections.get_credentials(&connection_id);
    state.manager.connect(&connection_id, &config, credentials).await
}

/// Closes the driver and all sessions for a connection.
#[tauri::command]
pub async fn disconnect(state: State<'_, AppState>, connection_id: String) -> AppResult<()> {
    state.manager.disconnect(&connection_id).await
}

// --- Schema -------------------------------------------------------------------

#[tauri::command]
pub async fn list_databases(state: State<'_, AppState>, connection_id: String) -> AppResult<Vec<String>> {
    state.manager.driver(&connection_id)?.list_databases().await
}

#[tauri::command]
pub async fn list_tables(
    state: State<'_, AppState>,
    connection_id: String,
    database: String,
) -> AppResult<Vec<TableInfo>> {
    state.manager.driver(&connection_id)?.list_tables(&database).await
}

#[tauri::command]
pub async fn list_columns(
    state: State<'_, AppState>,
    connection_id: String,
    database: String,
    table: String,
) -> AppResult<Vec<ColumnInfo>> {
    state
        .manager
        .driver(&connection_id)?
        .list_columns(&database, &table)
        .await
}

#[tauri::command]
pub async fn list_indexes(
    state: State<'_, AppState>,
    connection_id: String,
    database: String,
    table: String,
) -> AppResult<Vec<IndexInfo>> {
    state
        .manager
        .driver(&connection_id)?
        .list_indexes(&database, &table)
        .await
}

#[tauri::command]
pub async fn list_foreign_keys(
    state: State<'_, AppState>,
    connection_id: String,
    database: String,
    table: String,
) -> AppResult<Vec<ForeignKeyInfo>> {
    state
        .manager
        .driver(&connection_id)?
        .list_foreign_keys(&database, &table)
        .await
}

#[tauri::command]
pub async fn get_table_ddl(
    state: State<'_, AppState>,
    connection_id: String,
    database: String,
    table: String,
) -> AppResult<String> {
    state.manager.driver(&connection_id)?.table_ddl(&database, &table).await
}

// --- Execution ----------------------------------------------------------------

/// Splits sql into statements and executes them sequentially in the session.
#[tauri::command]
pub async fn execute_query(state: State<'_, AppState>, request: ExecuteRequest) -> AppResult<Vec<StatementResult>> {
    execute::execute(&state.manager, &state.history, request).await
}

/// Re-runs a statement without the row limit and writes the whole result set to a file.
#[tauri::command]
pub async fn export_query(state: State<'_, AppState>, request: ExportRequest) -> AppResult<ExportSummary> {
    execute::export(&state.manager, request).await
}

/// Writes rows the frontend already holds (the grid contents) to a file.
#[tauri::command]
pub async fn export_rows(request: RowsExportRequest) -> AppResult<ExportSummary> {
    tokio::task::spawn_blocking(move || export::export_rows(request))
        .await
        .map_err(|e| crate::error::AppError::Other(e.to_string()))?
}

/// Cancels the statement started with this queryId (KILL QUERY, pg_cancel, ...).
#[tauri::command]
pub async fn cancel_query(state: State<'_, AppState>, connection_id: String, query_id: String) -> AppResult<()> {
    state.manager.cancel_query(&connection_id, &query_id).await
}

/// Executes parameterized statements in a single transaction (data editing).
#[tauri::command]
pub async fn apply_changes(
    state: State<'_, AppState>,
    connection_id: String,
    session_id: String,
    statements: Vec<ParamStatement>,
) -> AppResult<ApplyResult> {
    execute::apply_changes(&state.manager, &connection_id, &session_id, statements).await
}

/// Closes the session connection (when a tab is closed).
#[tauri::command]
pub async fn close_session(state: State<'_, AppState>, connection_id: String, session_id: String) -> AppResult<()> {
    state.manager.close_session(&connection_id, &session_id).await
}

// --- History -----------------------------------------------------------------

#[tauri::command]
pub async fn list_history(state: State<'_, AppState>, limit: usize) -> AppResult<Vec<QueryHistoryEntry>> {
    Ok(state.history.list(limit))
}

#[tauri::command]
pub async fn clear_history(state: State<'_, AppState>) -> AppResult<()> {
    state.history.clear()
}
